src/timemachine-heap.c

```diff
@@ -0,0 +1,150 @@
+    if (unlikely(heap == NULL)) {
+        return NULL;     
+    }
+
+    heap->conf = conf;
+    heap->used_pool_count = 0;
+    heap->unused_pool_count = 0;
+    
+    TAILQ_INIT(&heap->used_mem_pools);
+    TAILQ_INIT(&heap->unused_mem_pools);
+    TAILQ_INIT(&heap->unused_packets);    
+    
+    uint32_t expanded = TimeMachineHeapExpand(tmtv, heap, conf->prealloc_count);
+    if (expanded != conf->prealloc_count) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Error: Could not allocate initial memory for "
+                   "time machine, expanded heap by: %" PRIu32 ", requested: %" PRIu32 
+                   " current Memory: %" PRIu64 " max Memory: %" PRIu64, expanded, 
+                   conf->prealloc_count, tmtv->current_memory, timemachine_config.max_memory);
+        exit(EXIT_FAILURE);                         
+    }
+    
+    return heap;
+}
+
+void TimeMachineHeapDestroy(TimeMachineHeap* heap) {
+    
+    if (heap == NULL) {
+        return;
+    }
+    
+    /* delete all the used pool data */
+    while (heap->used_pool_count > 0) {
+        TimeMachineMemPool* mem_pool;
+            
+        mem_pool = TAILQ_FIRST(&heap->used_mem_pools);
+        TAILQ_REMOVE(&heap->used_mem_pools, mem_pool, next);
+        SCFree(mem_pool);
+
+        heap->used_pool_count--;    
+    }
+    
+    /* delete all the unused pool data */
+    while (heap->unused_pool_count > 0) {
+        TimeMachinePacket* packet;
+        TimeMachineMemPool* mem_pool;
+
+        packet = TAILQ_FIRST(&heap->unused_packets);
+        TAILQ_REMOVE(&heap->unused_packets, packet, next);
+        TimeMachinePacketDestroy(packet);
+            
+        mem_pool = TAILQ_FIRST(&heap->unused_mem_pools);
+        TAILQ_REMOVE(&heap->unused_mem_pools, mem_pool, next);
+        SCFree(mem_pool);
+                
+        heap->unused_pool_count--;
+    }    
+}
+
+uint32_t TimeMachineHeapExpand(TimeMachineThreadVars* tmtv, TimeMachineHeap* heap, uint32_t expand_by) {
+    uint32_t i;
+           
+    if (heap == NULL) {
+        return 0;
+    }
+
+    uint32_t expand_count = expand_by;
+    uint32_t expand_mem_size = sizeof(TimeMachineMemPool) + heap->conf->max_packet_size + 
+                      sizeof(TimeMachinePacket);
+                   
+    if ((expand_mem_size * expand_by) + tmtv->current_memory > timemachine_config.max_memory) {
+        expand_count = (timemachine_config.max_memory - tmtv->current_memory) / expand_mem_size + 1;
+    }    
+    
+    for (i = 0; i < expand_count; i++) {
+        TimeMachineMemPool* mem_pool;
+        TimeMachinePacket* packet;
+        
+        mem_pool = SCCalloc(sizeof(TimeMachineMemPool), 1);
+        if (mem_pool == NULL) {
+            return i;
+        }    
+        
+        mem_pool->mem = SCMalloc(heap->conf->max_packet_size);
+        if (mem_pool->mem == NULL) {
+            SCFree(mem_pool);
+            return i;
+        }
+        
+        packet = TimeMachinePacketNew();
```

Comment:
The TimeMachineHeap basically ensures ordering of data from oldest to newest  so you have item1->item2->item3->itemN.  

You still need something that maps those items in a logical ordering associated with a flow.  So in the case above, the flow may have item1 and item3 may be part of the same flow but item2 is part of a different flow.  This is where TimeMachinePacket comes in and ensures you have item1->item3->itemN mapping within a flow so when you want to access the data for the flow, there's a logical ordering from oldest to newest there as well.  It's super basic in the sense that at this point, I felt all you really need is the pcap header struct information containing the time + the actual packet data and a pointer to the next logical entity in the list for the flow.

In theory, you could just have another Pool that's the "TimeMachinePacketPool" that enqueues packets as they are being processed if TimeMachine is enabled.  This would hold the packets even after they are done processing.  You would still need to add another variable to the packet data structure that points to the next packet within the timemachine pool associated with the flow.  The downside is the timemachine is for memory efficiency and you would incur a decent amount of overhead by storing the extra packet information therefore you wouldn't be able to go as far back in time if you wanted to.  There would be an extra cost that I haven't looked into around managing/deleting the packets from the TimeMachinePacketPool instead of at the end of decoding.  The question I would have is whether storing the extra information within the packet data structure is useful to reference "in the past" if needed?